### DISCUSSION TITLE
How does reven work to log all CPU instructions

### DISCUSSION THREAD

**Discussion by tag5** [tetrane/reven-community#25] [General]
Hello,

There is something i do not understand: When we record a scenario with Reven, the full machine runs on qemu and reven is recording all cpu instructions. How does it work ? Is there a custom plugin in qemu or is there a native instrumentation mechanism in qemu ?

Thanks

**Comment by tag5** [comment]
Okay i've found: you work with panda-re.
But how do you record instructions for VirtualBox ?

**Comment by qb1** [comment]
Hello,

You're right, we're moslty using panda as a record / replay stack.

The VirtualBox record / replay is our legacy stack. Its pricinples are rather similar: it records non-deterministic events, and use them during the replay phase. The main difference is that it plugs onto the `VMENTER` / `VMEXIT` events during the record phase, and then leverages Bochs for the replay. The [documentation](https://doc.tetrane.com/enterprise/latest/Misc/Vbox-VM.html) gives some details about the limitations.

If you're curious, you can find our open-source repos here: https://github.com/tetrane/tetrane-oss

Mathieu.

**Reply by tag5** [reply]
Thanks a lot for your anwser.
